Only set the Page Padding via SafeAreaInset on iOS

PureWeen opened the pull request:
### Description of Change ###
-If platform isn't iOS then don't do anything when the user changes the SafeAreaPadding setting
-Check to see if the user has changed the padding when setting SafeAreaPadding to false so it doesn't overwrite what the user used.  



### Issues Resolved ### 
- fixes #3809 
- fixes #3866

### Platforms Affected ### 
- Core/XAML (all platforms)
- iOS
- Android
- UWP

### Before/After Screenshots ### 
<!-- If possible, take a screenshot of your test case before these changes were made and another screenshot after the changes were made to show possible visual changes. -->

Not applicable

### Testing Procedure ###
-For the issue created click the two buttons provided to set the padding and change whether Safe Area has been enabled. Ensure that the page padding reacts as you'd expect it to 

### PR Checklist ###

- [x] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

PureWeen commented:
build --uitests

StephaneDelcroix commented on Xamarin.Forms.Core/PlatformConfiguration/iOSSpecific/Page.cs:
```diff
@@ -60,10 +60,14 @@ public static UIStatusBarAnimation PreferredStatusBarUpdateAnimation(this IPlatf
 
 		public static readonly BindableProperty UseSafeAreaProperty = BindableProperty.Create("UseSafeArea", typeof(bool), typeof(Page), false, propertyChanged: (bindable, oldValue, newValue) =>
 		{
+			if (Device.RuntimePlatform != Device.iOS) return;
```
if you ever need to change this PR for another reason, it'd be a good opportunity to move the propertychangedhandler into a real method, for clarity

StephaneDelcroix commented on Xamarin.Forms.Core/PlatformConfiguration/iOSSpecific/Page.cs:
```diff
@@ -60,10 +60,14 @@ public static UIStatusBarAnimation PreferredStatusBarUpdateAnimation(this IPlatf
-				page.Padding = default(Thickness);
+				var safeAreaInsets = GetSafeAreaInsets(page);
+				if (safeAreaInsets == page.Padding)
+					page.Padding = default(Thickness);
```
not related to your change, but directly setting the `Padding` from the SafeAreaInsets looks a bad idea. we probably need something like an `ActualPadding` property that sum the padding with the insets

rmarinho replied:
Agree, i think what we decided at the time was like if you are setting padding you just don't use SafeAreInsets, if you are using padding you will be responsabile to make sure on iOS to you get the insets and add that t your padding. that's why we expose the Insets by itself also.

PureWeen replied:
Agree, I figured from the implementation that it was setup to be all or nothing and not additive. My other thought is that we are going to have to redo this for the Android notch at which point I feel like this code will just move directly into the APIs of the Forms controls opposed to being a platform thing at which point we could make all this a little bit better.

For example detecting rotation and things like that

PureWeen commented:
build --uitests

PureWeen commented:
@rmarinho  and @StephaneDelcroix  

I changed this one a bit more so if you could review again.  iOS 10 was also getting blanked out in the same way as Android which is why the iOS 10 tests were failing. I updated the logic to just ignore inset logic if not on ios 11 or newer.     I also just moved the logic down into the iOS PageRenderer as it's just easier to make more informed decisions from there.

StephaneDelcroix commented on Xamarin.Forms.Platform.iOS/Renderers/PageRenderer.cs:
```diff
@@ -215,6 +223,22 @@ public override UIKit.UIStatusBarAnimation PreferredStatusBarUpdateAnimation
+			{
+				var safeAreaInsets = SafeAreaInsets;
+				if (safeAreaInsets == Page.Padding)
+					Page.Padding = default(Thickness);
```
setting the `Padding` property from the renderer is even worse than the original version

PureWeen replied:
Why? This is being triggered only when that property is changed so it's functionally identical. 

Plus it's being set from inside the context of iOS so it has more accurate platform context. When it was being set inside shared code then we had to add if statements to make sure it wasn't effecting other platforms.  Which seems more inelegant to me.

samhouts replied:
@StephaneDelcroix Do you agree with @PureWeen? Let us know how to proceed! Thanks!

A1b2c3d4e55 commented:
I’m not sure what time but I’ll be there 



Sent from Mail.Ru app for iOS


Friday, September 21, 2018, 10:15 AM +0300 from notifications@github.com <notifications@github.com>:
>@StephaneDelcroix  commented on this pull request.
>----------------------------------------------------------------------
>In  Xamarin.Forms.Platform.iOS/Renderers/PageRenderer.cs :
>> @@ -215,6 +223,22 @@ public override UIKit.UIStatusBarAnimation PreferredStatusBarUpdateAnimation
> 			}
> 		}
> 
>+		void UpdateUseSafeArea()
>+		{
>+			if (!Forms.IsiOS11OrNewer) return;
>+
>+			if (!UsingSafeArea)
>+			{
>+				var safeAreaInsets = SafeAreaInsets;
>+				if (safeAreaInsets == Page.Padding)
>+					Page.Padding = default(Thickness);
>
>setting the Padding property from the renderer is even worse than the original version
>—
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly,  view it on GitHub , or  mute the thread .

samhouts commented:
@rui Please re-review. Thanks!

PureWeen commented:
@rmarinho & @StephaneDelcroix  
I also added some additional breaking change notes which might need to be fixed or changed

StephaneDelcroix commented:
@samhouts @PureWeen I would:
- do the minimal change to fix the issue (check for iOS plat, trigger update on property changes)
- get this merged

and then, 
- deprecate the `UseSafeArea` property, only keeping the `SafeAreaInsets` one, allowing the user to bind Padding to the former

despite the flaws (not inherent to this PR, but the original design choices), I already approved this PR, indicating that this is not the hill I plan to die on

PureWeen commented:
@samhouts @StephaneDelcroix 

So, I was mistaken about this being breaking. It's really a fix of a regression in 3.2 and not a breaking change

3.2 makes it so the ordering of how these are specified in the XAML can effect ios10

for example
```xaml
			 Padding="25,25,25,25"
			ios:Page.UseSafeArea="true"
```

Will cause iOS 10 to set its padding to 0,0,0,0 where as
```xaml
			ios:Page.UseSafeArea="true"
			 Padding="25,25,25,25"
```

Will cause iOS 10 to set its padding to 25,25,25,25
In XF 3.1 both of these scenarios would have resulted in 25,25,25,25 for the padding

With the changes I've made in this PR it makes it so that the behavior works as it did on 3.1

So I still think what I have is the best way to maintain 3.1 behavior while maintaining the changes that @rui made initially that caused the regression but fixed a different issue

rmarinho commented:
👍 ffrom me

samhouts commented:
While we're in here, let's make sure we're not missing any null checks. https://github.com/xamarin/Xamarin.Forms/issues/3866

PureWeen commented:
@samhouts  thanks for the reminder.. I'll see if I can recreate what they are seeing and include a fix if it already hasn't been fixed

samhouts commented:
@StephaneDelcroix @PureWeen So we don't forget that this API needs work, I opened https://github.com/xamarin/Xamarin.Forms/issues/3965. For now, I think we should merge this to fix this issue. @StephaneDelcroix objections?

StephaneDelcroix commented:
@samhouts well aware of the current design flaws, I approved this PR 2 weeks ago, I'm not going to change my mind :)
